Key mapping feature

telday opened the pull request:
In the name of moving toward a more event driven interface, and having greater customization I have added KeyMaps.

### Why
The basic theory is that each mapped key corresponds either to a method/function, or to another mapped key which then corresponds to a method. The original intent was twofold, to make declaring new keybindings simpler/to cleanup the handle_key_press methods, as well as making it easier to have multiple binds for the same action. I wanted to include some vim-like movement in my TUI, Now this can be done simply with four lines of code, an if you were to make a KeyMap and store it in a config file it could be used in multiple different programs.

### Changes
The biggest change is the introduction of the KeyMap class, which provides the main functionality for the new mapping. Every widget/popup has a key_map which it inherits from the base Widget/Popup class. The children can then simply add their bindings to the parent key_map and when the parent goes to handle a key_press it will automatically consider the children's bindings.

### Considerations
I have not fully finished implementing this feature yet, I will need to write a few docstrings and upate tests/examples to reflect the new api. I also feel like the scope of the change would warrant a MINOR version increment to 0.2.0 rather than PATCH

Let me know what you think, I have a few ideas about using decorators to perform the same duty as the KeyMap.bind_key method currently does. This would increase the event driven nature and reduce some excess code.

I have made an updated verision of the todo-list example with the new key binding method and the vim controls [here](https://gist.github.com/telday/227d39502d2fe017deb8b8b3c3b007f9)

telday committed:
Added keymap class and make keys an enum

telday committed:
Merge branch 'master' into key_mapping

telday committed:
update gitignore

telday committed:
Changed key names

telday committed:
Added __add__ to keymap

telday committed:
Added new key_map to PyCUI

telday committed:
Added key maps to widgets

telday committed:
Moved popups to key map

telday committed:
Fixed syntax error

telday committed:
Finished basic implementation of key_map

telday committed:
Updated authors

jwlodek commented:
I like these changes in theory, but I wish they didn't break compatibility with previous versions. Is there any way we can keep the `add_key_command` function that just adds to the `raw_key_map`, and keep variables that reference the keys with the old naming scheme?

telday committed:
Update .gitignore

telday committed:
Updated docs

telday committed:
Changed tests to import new key enum

telday committed:
Added decorator to let methods which dont need to know their key ignore

telday committed:
Downgraded method call

telday committed:
Removed string formatters to comply with python3.5

telday committed:
Updated examples to match new api

telday commented:
I added back in the `add_key_command` method, which simply acts as a wrapper for making a new key_bind.

I removed some of the features used in order to comply with python 3.5. However I cannot get it actually working on windows because of windows-curses issues. I think it might be worth thinking about dropping 3.5 support.

I am working on a few small fixes on the examples right now then this should be okay to merge

telday committed:
Updated multi-window example

telday committed:
Re-added add_key_command and fixed issue with window closing

telday committed:
Updated autogit example

telday commented:
Everything looks good now, I ported over all of the examples, however did not change them to directly using the `key_map` in place of `add_key_command`. I tried to test everything to the best of my abilities, and it seems good on both Fedora 32 and Windows 10. Let me know if you find any issues or if there is something you think should change

jwlodek commented:
I'll try and get to this pull request this week, I've been very busy lately, so not much time for fun projects. I want to merge it, but I would like to test everything on my end, and check how much work it would be to convert an existing project like `pyautogit` to use this, I just haven't had time to.
